File: python/range_sum_query_immutable.py
# solution 1: compute the sum needed when calling `sumRange`
class NumArray:

    def __init__(self, nums: List[int]):
        self.nums = nums

    def sumRange(self, left: int, right: int) -> int:
        return sum(self.nums[left:right+1])


# solution 2: compute the sums of [0..i] on construction
class NumArray:

    def __init__(self, nums: List[int]):
        # could make this more efficient with `nwise`
        self.nums = [0 for _ in range(len(nums))]
        self.nums[0] = nums[0]
        for i in range(1, len(nums)):
            self.nums[i] = self.nums[i-1] + nums[i]

    def sumRange(self, left: int, right: int) -> int:
        if left > right:
            raise ValueError(f"left bound {left} greater than right bound {right}")
        if left == 0:
            return self.nums[right]
        return self.nums[right] - self.nums[left-1]


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)

File: rust/src/range_sum_query_immutable.rs
struct NumArray {
    range: Vec<i32>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl NumArray {

    fn new(nums: Vec<i32>) -> Self {
        let mut iter = nums.iter().scan(0, |acc, &n| {
            *acc = *acc + n;
            Some(*acc)
        });
        return NumArray {range: iter.collect()};
    }

    fn sum_range(&self, left: i32, right: i32) -> i32 {
        if left > right {
            panic!(
                format!("left bound must be strictly less than right bound but sum_range was called with {0} and {1}", left, right)
            );
        }
        if left == 0 {
            return self.range[right as usize];
        }
        return self.range[right as usize] - self.range[(left-1) as usize];
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * let obj = NumArray::new(nums);
 * let ret_1: i32 = obj.sum_range(left, right);
 */
